File: C-Plus-Plus/cp/tower_of_hanoi.cpp
/* Tower Of Hanoi
Tower of Hanoi is a mathematical puzzle where we have three rods and n disks. The objective of the puzzle is to move the entire stack to another rod using the below rules:
*Only one disk can be moved at a time.
*Each move consists of taking the upper disk from one of the stacks and placing it on top of another stack i.e. a disk can only be moved if it is the uppermost disk on a stack.
*No disk must be placed on top of a smaller disk.
*/
#include <bits/stdc++.h>
using namespace std;
void towerOfHanoi(int n, char from_rod, char to_rod, char intermediate_rod)
{
    if (n == 1)
    {
        cout << "Move disk 1 from rod " << from_rod << " to rod " << to_rod << endl;
        return;
    }
    towerOfHanoi(n - 1, from_rod, intermediate_rod, to_rod);
    cout << "Move disk " << n << " from rod " << from_rod << " to rod " << to_rod << endl;
    towerOfHanoi(n - 1, intermediate_rod, to_rod, from_rod);
}
int main()
{
    int n;
    cin >> n;
    towerOfHanoi(n, 'A', 'C', 'B');
}
/*
Time Complexity: O(n)
INPUT:
4
OUTPUT:
Move disk 1 from rod A to rod B
Move disk 2 from rod A to rod C
Move disk 1 from rod B to rod C
Move disk 3 from rod A to rod B
Move disk 1 from rod C to rod A
Move disk 2 from rod C to rod B
Move disk 1 from rod A to rod B
Move disk 4 from rod A to rod C
Move disk 1 from rod B to rod C
Move disk 2 from rod B to rod A
Move disk 1 from rod C to rod A
Move disk 3 from rod B to rod C
Move disk 1 from rod A to rod B
Move disk 2 from rod A to rod C
Move disk 1 from rod B to rod C
*/

File: Go/other/tower_of_hanoi/Tower_of_Hanoi.go
package main

import "fmt"

type solver interface {
	play(int)
}

type towers struct{}

func (t *towers) play(n int) {
	t.shiftN(n, 1, 2, 3)
}

func (t *towers) shiftN(n, from, to, via int) {
	if n > 0 {
		t.shiftN(n-1, from, via, to)
		t.shiftM(from, to)
		t.shiftN(n-1, via, to, from)
	}
}

func (t *towers) shiftM(from, to int) {
	fmt.Println("Shifting disk from rod", from, "to rod", to)
}

func main() {
	var t solver
	t = new(towers)
	t.play(4)
}

/*
Tower of Hanoi is a mathematical puzzle where we have three rods and n disks. The objective of the puzzle is to move the entire stack to another rod, obeying the following simple rules:
1) Only one disk can be moved at a time.
2) Each move consists of taking the upper disk from one of the stacks and placing it on top of another stack i.e. a disk can only be moved if it is the uppermost disk on a stack.
3) No disk may be placed on top of a smaller disk.
In this code, we use the recursive definition of Tower of Hanoi.
Time Complexity: O(2^N)
Space Complexity: O(N)

INPUT: 4

OUTPUT:
Shifting disk from rod 1 to rod 3
Shifting disk from rod 1 to rod 2
Shifting disk from rod 3 to rod 2
Shifting disk from rod 1 to rod 3
Shifting disk from rod 2 to rod 1
Shifting disk from rod 2 to rod 3
Shifting disk from rod 1 to rod 3
Shifting disk from rod 1 to rod 2
Shifting disk from rod 3 to rod 2
Shifting disk from rod 3 to rod 1
Shifting disk from rod 2 to rod 1
Shifting disk from rod 3 to rod 2
Shifting disk from rod 1 to rod 3
Shifting disk from rod 1 to rod 2
Shifting disk from rod 3 to rod 2

*/
